src/main/kotlin/com/terraformation/backend/nursery/db/BatchImporter.kt

```diff
@@ -0,0 +1,126 @@
+import com.terraformation.backend.db.nursery.tables.pojos.BatchesRow
+import com.terraformation.backend.file.FileStore
+import com.terraformation.backend.file.UploadService
+import com.terraformation.backend.file.UploadStore
+import com.terraformation.backend.i18n.Messages
+import com.terraformation.backend.importer.CsvImporter
+import com.terraformation.backend.importer.CsvValidator
+import com.terraformation.backend.species.db.SpeciesStore
+import java.io.InputStream
+import java.time.LocalDate
+import javax.annotation.ManagedBean
+import org.jobrunr.scheduling.JobScheduler
+import org.jooq.DSLContext
+import org.springframework.context.annotation.Lazy
+
+@ManagedBean
+class BatchImporter(
+    private val batchStore: BatchStore,
+    dslContext: DSLContext,
+    fileStore: FileStore,
+    private val messages: Messages,
+    private val parentStore: ParentStore,
+    @Lazy scheduler: JobScheduler,
+    private val speciesStore: SpeciesStore,
+    uploadProblemsDao: UploadProblemsDao,
+    uploadsDao: UploadsDao,
+    uploadService: UploadService,
+    uploadStore: UploadStore,
+    userStore: UserStore,
+) :
+    CsvImporter(
+        dslContext,
+        fileStore,
+        scheduler,
+        uploadProblemsDao,
+        uploadsDao,
+        uploadService,
+        uploadStore,
+        userStore) {
+  override val templatePath: String
+    get() = "/csv//batches-template.csv"
+
+  fun receiveCsv(inputStream: InputStream, fileName: String, facilityId: FacilityId): UploadId {
+    requirePermissions { createBatch(facilityId) }
+
+    if (parentStore.getFacilityType(facilityId) != FacilityType.Nursery) {
+      throw FacilityTypeMismatchException(facilityId, FacilityType.Nursery)
+    }
+
+    val organizationId = parentStore.getOrganizationId(facilityId)
+    return doReceiveCsv(
+        inputStream, fileName, UploadType.SeedlingBatchCSV, organizationId, facilityId)
+  }
+
+  override fun getValidator(uploadsRow: UploadsRow): CsvValidator {
+    return BatchCsvValidator(uploadsRow.id!!, messages)
+  }
+
+  override fun doImportCsv(
+      uploadsRow: UploadsRow,
+      csvReader: CSVReader,
+      overwriteExisting: Boolean
+  ) {
+    val organizationId = uploadsRow.organizationId!!
+    val facilityId = uploadsRow.facilityId!!
+
+    requirePermissions {
+      createBatch(facilityId)
+      createSpecies(organizationId)
+    }
+
+    // Consume header row
+    csvReader.readNext()
+
+    dslContext.transaction { _ -> csvReader.forEach { importRow(it, organizationId, facilityId) } }
+  }
+
+  private fun importRow(
+      rawValues: Array<String?>,
+      organizationId: OrganizationId,
+      facilityId: FacilityId
+  ) {
+    val values = rawValues.map { it?.trim()?.ifEmpty { null } }
+
+    val scientificName = values[0]!!
+    val commonName = values[1]
+    val germinatingQuantity = values[2]?.toInt() ?: 0
+    val seedlingQuantity = values[3]?.toInt() ?: 0
+    val storedDate = LocalDate.parse(values[4])!!
+
+    val speciesId =
```

Comment:
If i understand correctly, this line creates a species row in the DB if it doesn't exist before?

Reply:
> If i understand correctly, this line creates a species row in the DB if it doesn't exist before?

Yes, that's correct.